### PR TITLE
To do list

### PR DESCRIPTION
**Pull request by gosiaOsojca**

### CODE REVIEW

File: public/manifest.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (kononenko-a-m) [inline public/manifest.json:1]: can you please remove `public/mainfest.json`, `public/logo192.png`, `public/logo512.png` files as you don't need them?

File: src/App.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import { Component } from 'react';
+import AppHeader from './AppHeader';
+import ItemsList from './ItemsList';
+import AppFooter from './AppFooter';
+
+class App extends Component {
+    constructor() {
+        super()
+        this.state = {
+            inputfield: '',
```
Comment (kononenko-a-m) [inline src/App.js]: 2 things:
1. Please follow one naming convention, here you have `inputfield` (all lower case characters) and below you have `saveButtonClassActive` (camelCase).
2. Please, try to avoid referring DOM in the state names. So instead of `inputfield`, `newToDoItem` would be much more preferred.

File: src/App.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            inputs: [],
```
Comment (kononenko-a-m) [inline src/App.js]: the same here about avoiding referring DOM in the state names. Preferred name can be `toDoItems`

File: src/App.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            saveButtonClassActive: false
```
Comment (kononenko-a-m) [inline src/App.js]: the same here about avoiding referring DOM in the state names. `persistData` can be better name here

File: src/App.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        };
+    }
+
+    componentDidMount() {
+        const storageItems = localStorage.getItem('items');
+        const parsedItems = JSON.parse(storageItems);
+        if (parsedItems) {
+            const items = [];
```
Comment (kononenko-a-m) [inline src/App.js]: can you please change this to reduce?

File: src/App.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        super()
+        this.state = {
+            inputfield: '',
+            inputs: [],
+            saveButtonClassActive: false
+        };
+    }
+
+    componentDidMount() {
+        const storageItems = localStorage.getItem('items');
```
Comment (kononenko-a-m) [inline src/App.js]: can you please put 'items' in some `const` and use it here and in the place where you save it?

File: src/App.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        const parsedItems = JSON.parse(storageItems);
```
Comment (kononenko-a-m) [inline src/App.js]: `JSON.parse` may throw an error if there would be an invalid JSON in localStorage. I know, that you may think that how there can get some invalid JSON, but in reality, you may use some libraries which store data in localStorage, or the user may have some extension that writes data to the localStorage.

You should always take care of the case where there some invalid (or no) data in localStorage.

Please wrap this in try-catch and ensure that invalid data would be handled properly.

File: src/App.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        }
+    }
+
+    onInputChange = event => {
+        this.setState({ inputfield: event.target.value })
+    }
+
+    onAddButtonClick = () => {
+        if (this.state.inputfield) {
+            this.setState(state => {
```
Comment (kononenko-a-m) [inline src/App.js]: why are you using here function for setting a state?

I believe that this state update can be done without a state update function. Without this function, code would look a little bit simpler which makes this option as the preferred one.

File: src/App.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                    value: state.inputfield,
+                    isChecked: false
+                };
+                const inputs = state.inputs.concat(item);
+                return {
+                    inputs,
+                    inputfield: '',
+                };
+            }, () => {
+                if (this.state.saveButtonClassActive) {
```
Comment (kononenko-a-m) [inline src/App.js]: I believe updating value in localStorage should be more reactive, which means moving this logic inside `componentDidUpdate`, and I don't see any real reason to have it here.

File: src/App.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                    this.saveToLocalStorage();
+                }
+            })
+        } else {
+            alert("Fill the field");
```
Comment (kononenko-a-m) [inline src/App.js]: it doesn't look like proper error handling, please implement nice UI for handling scenario when field is empty

File: src/App.js [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+    }
+
+    onDeleteButtonClick = event => {
+        this.setState({
+            inputs: this.state.inputs.filter(function (item) {
+                if (item.value !== event.target.previousSibling.textContent) {
```
Comment (kononenko-a-m) [inline src/App.js]: please don't rely on `event.target.previousSibling.textContent` here.

The first reason for this, is that in general when you updating react state you should rely on the state itself or other logical objects that you introduce yourself and not DOM structure.

The second reason here would be that you create a tight implicit dependency between this function and where your delete button is placed. Imagine that you decide to add one more div between the delete button and the todo item. The change would look trivial however that change would lead to breaking this logic. In general that the real reason for the first reason 😊

Comment (kononenko-a-m) [reply src/App.js]: so, idea is that you should pass as parameter to this function not an event, but an item which user want to delete

File: src/App.js [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+                    return item;
+                }
+            })
+        }, () => {
+            if (this.state.saveButtonClassActive) {
```
Comment (kononenko-a-m) [inline src/App.js]: the same as above, this logic shouldn't be here.

and you actually duplicate logic here.

File: src/App.js [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    itemToggleClass = item => {
+        const items = this.state.inputs.map((input) => {
+            return item === input ? { ...item, isChecked: !item.isChecked } : input;
+        })
+
+        this.setState({ inputs: items });
+    }
+
+    saveToLocalStorage = () => {
+        localStorage.setItem('items', JSON.stringify(this.state.inputs));
```
Comment (kononenko-a-m) [inline src/App.js]: according to the MDN https://developer.mozilla.org/en-US/docs/Web/API/Storage/setItem#exceptions
`setItem` may throw an error, so please add handling for that case

File: src/App.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    removeFromLocalStorage = () => {
+        localStorage.removeItem('items');
+    }
+
+    render() {
+        const saveButtonClassActive = this.state.saveButtonClassActive;
+        if (saveButtonClassActive) {
```
Comment (kononenko-a-m) [inline src/App.js]: I believe you just need to remove it

File: src/AppFooter.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+function AppFooter({ changeButton, buttonClass, addToStorage}) {
```
Comment (kononenko-a-m) [inline src/AppFooter.js]: I don't think that your AppFooter should be aware of `addToStorage`, it should take 1 prop - onButtonChange

Comment (kononenko-a-m) [reply src/AppFooter.js]: there may be a bigger refactoring required with this component, however let's keep it for the future

File: src/AppFooter.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+    return (
+        <div className='footer'>
+            <button
+                className={buttonClass}
+                id='saveButton'
```
Comment (kononenko-a-m) [inline src/AppFooter.js]: it's a bad idea to use hardcoded html id's inside of React component, please don't use it

File: src/AppHeader.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+                    className='input'
+                    type='text'
+                    id='input'
+                    placeholder='New Task...'
+                    onChange={inputChange}
+                    value={inputValue}
+                />
+                <span
+                    className='add-button'
+                    id='addButton'
```
Comment (kononenko-a-m) [inline src/AppHeader.js]: again, please avoid id here and in input above

File: src/AppHeader.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+            <div className='header__input'>
+                <input
+                    className='input'
+                    type='text'
+                    id='input'
+                    placeholder='New Task...'
+                    onChange={inputChange}
+                    value={inputValue}
+                />
+                <span
```
Comment (kononenko-a-m) [inline src/AppHeader.js]: in your case, it's behave like a button, but you use `span` here, this is wrong from semantic HTML perpective

File: src/AppHeader.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+                    className='add-button'
+                    id='addButton'
+                    onClick={addItem}
```
Comment (kononenko-a-m) [inline src/AppHeader.js]: here you add and onClick, however there is a popular pattern on the web submit value with pressing Enter which you not supporting here.

I believe the easiest way for you to achieve it would be wrapping this button and input above in `form`, add to `form` onSubmit handler where you would make `e.preventDefault()` which would help you to avoid page refresh and then you call `addItem`. Also you need to add a button type="submit"

File: src/CloseButton.js [hunk 20]
```diff
@@ -0,0 +1,10 @@
+function CloseButton({removeItem}) {
+    return (
+        <span
```
Comment (kononenko-a-m) [inline src/CloseButton.js]: again, this behave as button, but right now it's a span, which is wrong, please change it to proper HTML element

File: src/ItemsList.js [hunk 21]
```diff
@@ -0,0 +1,23 @@
+import Item from './Item';
+
+function ItemsList({ items, removeItem, changeItemStatus }) {
+    return (
+        <div className='content'>
+            <ul className='list' id='list'>
+                {items.map(item => {
+                    return (
+                        <Item
+                            className={`list__item` + (item.isChecked ? ' list__item--checked' : ' rhyryh')}
```
Comment (kononenko-a-m) [inline src/ItemsList.js]: what `rhyryh` means?

File: src/ItemsList.js [hunk 22]
```diff
@@ -0,0 +1,23 @@
+import Item from './Item';
+
+function ItemsList({ items, removeItem, changeItemStatus }) {
+    return (
+        <div className='content'>
+            <ul className='list' id='list'>
```
Comment (kononenko-a-m) [inline src/ItemsList.js]: again, please avoid ids

**Review by kononenko-a-m — CHANGES_REQUESTED** [review]

File: public/manifest.json (hunk 1 above)
Comment (gosiaOsojca) [reply public/manifest.json:1]: Ok, I have removed them.

File: src/App.js (hunk 2 above)
Comment (gosiaOsojca) [reply src/App.js]: Ok, I changed `inputfield` to `newToDoItem`

File: src/App.js (hunk 3 above)
Comment (gosiaOsojca) [reply src/App.js]: Sure, changed `inputs` to `toDoItems`.

File: src/App.js (hunk 4 above)
Comment (gosiaOsojca) [reply src/App.js]: Ok, changed `saveButtonClassActive` to `persistData`.

File: src/App.js (hunk 6 above)
Comment (gosiaOsojca) [reply src/App.js]: I created in class constructor `this.localStorageItems` to store the 'items'.

File: src/App.js (hunk 7 above)
Comment (gosiaOsojca) [reply src/App.js]: Ok, I have wrapped it in try-catch. But what should happen inside of catch? Just alert the error?

File: src/App.js (hunk 5 above)
Comment (gosiaOsojca) [reply src/App.js]: Sure, I changed the `for...of` to `reduce`. But why is `reduce` better here?

File: src/App.js (hunk 9 above)
Comment (gosiaOsojca) [reply src/App.js]: I moved this function to `componentDidUpdate`.

File: src/App.js (hunk 8 above)
Comment (gosiaOsojca) [reply src/App.js]: You're right. I can do this without the function as well.

File: src/App.js (hunk 12 above)
Comment (gosiaOsojca) [reply src/App.js]: Right. I have removed this part.

File: src/App.js (hunk 14 above)
Comment (gosiaOsojca) [reply src/App.js]: Right!

File: src/ItemsList.js (hunk 21 above)
Comment (gosiaOsojca) [reply src/ItemsList.js]: I used it for testing. I forgot to delete it.

File: src/ItemsList.js (hunk 22 above)
Comment (gosiaOsojca) [reply src/ItemsList.js]: Ok, removed.

File: src/AppHeader.js (hunk 17 above)
Comment (gosiaOsojca) [reply src/AppHeader.js]: Ok, removed.

File: src/AppFooter.js (hunk 16 above)
Comment (gosiaOsojca) [reply src/AppFooter.js]: Ok, I removed it.

File: src/App.js (hunk 11 above)
Comment (gosiaOsojca) [reply src/App.js]: Ok. I used here as parameter the item.

File: src/App.js (hunk 13 above)
Comment (gosiaOsojca) [reply src/App.js]: Ok, I have wrapped it in try-catch.

File: src/App.js (hunk 10 above)
Comment (gosiaOsojca) [reply src/App.js]: Right! I have added new styles for empty field.

File: src/AppHeader.js (hunk 18 above)
Comment (gosiaOsojca) [reply src/AppHeader.js]: Sure, I changed it to `button`.

File: src/AppHeader.js (hunk 19 above)
Comment (gosiaOsojca) [reply src/AppHeader.js]: I used here form and it works now fine.

File: src/CloseButton.js (hunk 20 above)
Comment (gosiaOsojca) [reply src/CloseButton.js]: I changed it to button.

File: src/App.js (hunk 7 above)
Comment (kononenko-a-m) [reply src/App.js]: that's a good question, common approach would be to put an error into the component state if occur and then show some nice UI message saying that something went wrong, and unfortunately we cannot restore saved todoItems

File: src/App.js (hunk 5 above)
Comment (kononenko-a-m) [reply src/App.js]: well, probably you're right, `.map` is more appropriate.

the reason I suggest `.reduce`, because I thought that you're reducing `parsedItems` into `items`, but you actually remapping it, so proper solution would be `.map`

File: src/App.js [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    removeFromLocalStorage = () => {
+        localStorage.removeItem(this.localStorageItems);
+    }
+
+    render() {
+        return (
+            <div className='container'>
+                <AppHeader inputClass={this.state.incorrectEntry ? 'input input--incorrect' : 'input'} inputChange={this.onInputChange} addItem={this.onAddButtonClick} inputValue={this.state.newToDoItem} errorMessageClass={this.state.incorrectEntry ? 'error-message error-message--visible' : 'error-message'}/>
```
Comment (kononenko-a-m) [inline src/App.js]: ok, now as you've cleaned up a little bit of logic, we can go to restructuring.

first thing, can you please create a new directory `modules` and inside of it `todo-list` directory?
after that can you please move out all logic from the app into a new component (you can name it like `TodoList`) inside of `modules/todo-list` directory?

next thing, inside of this component lift up logic from `AppHeader` and `AppFooter` to the new `TodoList` (this would lead to removal of `AppHeader` and `AppFooter`)

next thing, please add to the project react-router (https://reactrouter.com/) and please in App component add routing for `/` path which would render the new `TodoList` component

File: src/App.js (hunk 5 above)
Comment (gosiaOsojca) [reply src/App.js]: Ok, I changed it to `map`.

File: src/App.js (hunk 23 above)
Comment (gosiaOsojca) [reply src/App.js]: I'm not sure if I understand you well. Should I move everything from `App.js` into the new `ToDoList` component? So should I remove the `App.js`?

Comment (gosiaOsojca) [reply src/App.js]: Ok, I changed the structure and added the react-router.

File: src/App.js [hunk 24]
```diff
@@ -0,0 +1,127 @@
+import TodoList from './modules/todo-list/TodoList';
+import SignUp from './modules/SignUp';
+import SignIn from './modules/SignIn';
+import SignUpConfirmation from './modules/SignUpConfirmation';
+import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
+import { useState, useEffect } from 'react';
+
+
+function App() {
+    const [newToDoItem, setNewToDoItem] = useState('');
```
Comment (kononenko-a-m) [inline src/App.js]: please movo todo related logic to todo list component

File: src/App_old.js [hunk 25]
```diff
@@ -0,0 +1,123 @@
+import { Component } from 'react';
```
Comment (kononenko-a-m) [inline src/App_old.js]: if it's old, then please remove from the git

File: src/modules/SignIn.js [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+            password: values.password
+        }
+
+        const requestOptions = {
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify(jsonData)
+        };
+
+        fetch('http://localhost:8888/v1/user/sign-in', requestOptions)
```
Comment (kononenko-a-m) [inline src/modules/SignIn.js:37]: please handle when promise fail

Comment (kononenko-a-m) [inline src/modules/SignIn.js:37]: what if the user leaves the page when your request in progress?
how you think it's can be handled?

File: src/modules/SignIn.js [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+import FormButton from '../FormButton';
+
+
+function SignIn() {
+    const initialValues = {
+        email: '',
+        password: '',
+    };
+
+    let validate = (values) => {
```
Comment (kononenko-a-m) [inline src/modules/SignIn.js]: this looks good, however, maybe you could use Yup for validation?

File: src/modules/SignUp.js [hunk 28]
```diff
@@ -0,0 +1,150 @@
...
+
+
+function SignUp() {
+    const initialValues = {
+        name: '',
+        email: '',
+        password: '',
+    };
+
+    let validate = (values) => {
```
Comment (kononenko-a-m) [inline src/modules/SignUp.js]: the same, can you please Yup?

File: src/modules/SignUpConfirmation.js [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+        history.push('/SignIn');
+    }
+
+    return (
+        <>
+            <div className='header'>
+                <h2 className='header__title'>Congratulations!</h2>
+            </div>
+            <div class='sign-up-confirmation'>
+                <span className='sign-up-confirmation--span'>Your account has been successfully created.<br></br> Please continue to Log In.</span>
```
Comment (kononenko-a-m) [inline src/modules/SignUpConfirmation.js]: please, never use `<br>`
the only exception may be if you making email template

File: src/modules/SignUp.js [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+                    handleSubmit,
+                    errors,
+                    touched,
+                    handleBlur,
+                    isValid,
+                    dirty
+                } = formik;
+                return (
+                    <>
+                        <div className='header'>
```
Comment (kononenko-a-m) [inline src/modules/SignUp.js:66]: I believe you need to separate some SinglePanelLayout that encapsulate where and how to show header, and it's content

File: src/modules/SignUp.js [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+            password: values.password
+        }
+
+        const requestOptions = {
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify(jsonData)
+        };
+
+        fetch('http://localhost:8888/v1/user/sign-up', requestOptions)
```
Comment (kononenko-a-m) [inline src/modules/SignUp.js:45]: the same issues as for request above

Comment (kononenko-a-m) [inline src/modules/SignUp.js:45]: also maybe it worth creating a function like `apiCall()` and `postApiCall()` accordingly which can encode as JSON and add default headers

File: src/App.js [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+    };
+
+    return (
+        <Router>
+            <div className='container'>
+                <Switch>
+                    <Route exact path='/'>
+                        <SignUp />
+                    </Route>
+                    <Route path='/SignUpConfirmation'>
```
Comment (kononenko-a-m) [inline src/App.js]: please don't use upper case in routes

File: src/modules/SignIn.js [hunk 33]
```diff
@@ -0,0 +1,122 @@
+import { Formik } from "formik";
+import { useHistory } from "react-router-dom";
+import FormInput from '../FormInput';
+import FormButton from '../FormButton';
+
+
+function SignIn() {
```
Comment (kononenko-a-m) [inline src/modules/SignIn.js:6]: this and SignUp component have some validation logic and critical business logic as well
I think this is time to add tests for this done in `testing-library`. Please add tests for SignIn and SignUp components

File: src/App.js (hunk 24 above)
Comment (gosiaOsojca) [reply src/App.js]: Sure, I moved everything to `ToDoList`

File: src/App_old.js (hunk 25 above)
Comment (gosiaOsojca) [reply src/App_old.js]: Ok, I removed it.

File: src/App.js (hunk 32 above)
Comment (gosiaOsojca) [reply src/App.js]: Ok, I changed it

File: src/modules/SignUpConfirmation.js (hunk 29 above)
Comment (gosiaOsojca) [reply src/modules/SignUpConfirmation.js]: Ok, I removed `br`

File: src/modules/SignIn.js (hunk 26 above)
Comment (gosiaOsojca) [reply src/modules/SignIn.js:37]: Ok, I have added `.catch`

Comment (gosiaOsojca) [reply src/modules/SignIn.js:37]: Hmm, the request should be 'stopped' somehow?? So that the user has to log in again after returning to the page.

File: src/modules/SignUp.js (hunk 31 above)
Comment (gosiaOsojca) [reply src/modules/SignUp.js:45]: Right!

File: src/modules/SignUp.js (hunk 30 above)
Comment (gosiaOsojca) [reply src/modules/SignUp.js:66]: That means I have to create for te header a separate component and then use it in all app views?

File: src/modules/SignIn.js (hunk 27 above)
Comment (gosiaOsojca) [reply src/modules/SignIn.js]: Ok, I switched to Yup

File: src/modules/SignUp.js (hunk 28 above)
Comment (gosiaOsojca) [reply src/modules/SignUp.js]: Ok, Yup is there

File: src/App.js (hunk 24 above)
Comment (kononenko-a-m) [reply src/App.js]: have you pushed those changes?

File: src/modules/SignIn.js (hunk 26 above)
Comment (kononenko-a-m) [reply src/modules/SignIn.js:37]: ok, so if you want to implement a stop/cancel pattern then you can go with AbortController https://developer.mozilla.org/en-US/docs/Web/API/AbortController

however, you need to understand that it may not work as you think. for example, browser sent a request, the server started handling it, and then the browser cancels the request, here 2 things may happen: 

1. server may not have cancelation logic handling, so your cancellation may help only before server started handling the request, but after it wouldn't make any impact on the operation
2. server know how to handle cancellation logic, but cancel is actually triggered after the server already persisted data in DB, so it wouldn't do any change

in summary, a cancel request is not the best for creating/updating/deleting requests, however, in some cases, it still would be a valid pattern. a more common way to handle it is to block UI and try to prevent users from leaving the page or using an optimistic UI approach.

cancel requests is very helpful when you have read requests, which can be canceled without any problems.

File: src/App.js (hunk 24 above)
Comment (gosiaOsojca) [reply src/App.js]: No, not yet, I will push all changes together
